### PR TITLE
Fix false positive when testing for-loops for unbalanced unpacking (W0644)

### LINKED ISSUE #8156: False reports on W0644 unbalanced-dict-unpacking in pylint 2.16.0
### Bug description

When unpacking while iterating over dict.values() we seem to get a spurious warning about unbalanced dict unpacking.

Here's a test-case,  result vs. result1 may require a few changes to the types, but the original code was doing a lot more dynamic things with callbacks and more.

```python
from typing import Dict, Tuple, Callable

Inner = Callable[[int], Dict[str, int]]

def returns_dict(a:int, b:int) -> Dict[str, Tuple[int, int, int]]:
    result1 = {
        "abcdef": returns_things(a, b, "abcdef"),
        "cdef": returns_other_things(a,b, "cdef"),
    }
    result = {
        "abc": (1, 2, 3),
        "cde": (1, 2, 3),
    }
    return result

def returns_things(a:int, b:int, c:str) -> Tuple[str, str, Inner]:
    def inner(other: int) -> Dict[str, int]:
        return {"called": other}
    return f"{a}{b}", c, inner


def returns_other_things(a:int, b:int, c:str)  -> Tuple[str, str, Inner]:
    def inner(other: int) -> Dict[str, int]:
        return {"called": other}
    return f"{1}{c}", c, inner


def thing(a:int, b:int) -> None:
    bar = returns_dict(a, b)
    [print(x) for x in bar.values()]
    for x, y, _z in bar.values():
        assert x
        assert y

if __name__ == "__main__":
    thing(1, 2)

```

### Configuration

_No response_

### Command used

```shell
pylint pylint_testcase.py
```

### Pylint output

```shell
************* Module pylint_testcase
pylint_testcase.py:32:4: W0644: Possible unbalanced dict unpacking with bar.values(): left side has 3 labels, right side has 2 values (unbalanced-dict-unpacking)
```

### Expected behavior

No warning should happen

### Pylint version

```shell
pylint 2.16.0
astroid 2.14.1
Python 3.11.1 (main, Jan  6 2023, 00:00:00) [GCC 12.2.1 20221121 (Red Hat 12.2.1-4)]
```

### OS / Environment

Fedora 37, x86_64

### Additional dependencies

_No response_

### PR DESCRIPTION
**Pull request by Neowizard**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

When checking for-loops for unbalanced dict unpacking, Pylint would compare the length of the targets with the length of the values iterable. However, since the for-loop would iterate the values and unpack each value in turn, Pylint should compare the length of the targets with the length of each value (and not the total length of values).

Closes #8156

### CODE REVIEW

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Look amazing, great work ! It's probably going to close a good number of issues, Could you reset the doc upgrade, this is something we need to fix in another MR imo.

**Comment by Neowizard** [comment]
> Could you reset the doc upgrade, this is something we need to fix in another MR imo.

Yeah, did that by accident. Didn't mean to edit those files, but got confused with how to use `towncrier` and somehow ended up running some command that refactored them. Still not sure how did that, but it's reverted now.

**Comment by Neowizard** [comment]
@Pierre-Sassoulas, I reverted the docs changes (forgot to tag you on my previous comment), but I'm having trouble figuring out the issue with the `readthedocs` check. It's complaining about the format of fragment `2018.user_action`, but I'm not sure how I caused this. 

Do you have an idea what's the issue here?

**Comment by mbyrnepr2** [comment]
@Neowizard the ReadTheDocs issue is fixed on main now (I also ran into the same thing in a PR). Pull main and it should pass next time in the CI.

**Review by Pierre-Sassoulas — COMMENTED** [review]
LGTM, very nice change ! Just need a rebase, sorry for the issue with the documentation.

**Comment by Neowizard** [comment]
@Pierre-Sassoulas merged main into my branch, not all checks pass. Thanks.

File: tests/functional/d/dict_iter_missing_items.py
Comment (Pierre-Sassoulas) [inline tests/functional/d/dict_iter_missing_items.py:2]: Where is the new unbalanced-dict-unpacking ? (Iine 12 I suppose ?). We try to not raise 2 messages for the same problem but we might want to be explicit if we choose to do it here. Or we may not raise it in this specific case because we already raise a more accurate message. Maybe it's the occasion to merge the two checks and optimize the one we're merging in the other ?

```suggestion
```

**Review by Pierre-Sassoulas — COMMENTED** [review]
Thank you for rebasing, I did not check how much work merging those two message would take, (it's easier if they live in the same checkers). Let me know what you think is the best way forward :)

File: tests/functional/d/dict_iter_missing_items.py
Comment (pdbsvmct) [reply tests/functional/d/dict_iter_missing_items.py:2]: Yeah, line 12 yields that warning because we're unpacking the keys of a dict onto 2 variables.

As far as I can tell, there's no case where `dict-items-missing-iter` is raised and `unbalanced-dict-unpacking` isn't, since the check for missing `items` is implemented by looking for unbalanced dict unpacking - It looks for `for` loops with 2 targets while iterating a dict where keys aren't tuples. 

Maybe `dict-items-missing-iter` is deprecated at this point.

Comment (Neowizard) [reply tests/functional/d/dict_iter_missing_items.py:2]: Yeah, line 12 yields that warning because we're unpacking the keys of a dict onto 2 variables.

As far as I can tell, there's no case where `dict-items-missing-iter` is raised and `unbalanced-dict-unpacking` isn't, since the check for missing `items` is implemented by looking for unbalanced dict unpacking - It looks for `for` loops with 2 targets while iterating a dict where keys aren't tuples. 

Maybe `dict-items-missing-iter` is deprecated at this point.

Comment (Pierre-Sassoulas) [reply tests/functional/d/dict_iter_missing_items.py:2]: ``dict-items-missing-iter`` is a more precise message than ``unbalanced-dict-unpacking`` (it tells precisely what the solution is to fix the issue) so even if it would make sense for us to remove it to make our code leaner, I don't think that's the best solution from a user perspective.

Comment (Neowizard) [reply tests/functional/d/dict_iter_missing_items.py:2]: OK, yeah, that makes sense. I'll try and detect whether the conditions for `dict-items-missing-iter` hold when checking for `unbalanced-dict-unpacking` warning and omit the latter warning if they do.

Comment (Pierre-Sassoulas) [reply tests/functional/d/dict_iter_missing_items.py:2]: Sounds good !

Comment (pdbsvmct) [reply tests/functional/d/dict_iter_missing_items.py]: Implemented. Let me know if there's what you had in mind

**Review by Pierre-Sassoulas — APPROVED** [review]
LGTM, This is a pretty great first contribution 🔥 Let's wait a little for another maintainer opinion before merging but this is definitely going in the next 2.x release.

**Review by Pierre-Sassoulas — APPROVED** [review]

File: pylint/checkers/variables.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/variables.py:1333]:
```suggestion
            # TODO (performance), merging both checks would streamline this
```

**Review by Pierre-Sassoulas — APPROVED** [review]

File: pylint/checkers/variables.py
Comment (mbyrnepr2) [inline pylint/checkers/variables.py:3000]: There could be more cases like list comprehensions or set comprehensions:
```python
for x, y, z in {1: {i for i in [1,2,3, 4] if i >= 2}}.values():
    print(x)
```

File: pylint/checkers/variables.py
Comment (mbyrnepr2) [inline pylint/checkers/variables.py:1337]:
```suggestion
        is_starred_targets = any(isinstance(target, nodes.Starred) for target in targets)
```

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Note to self: find out if it closes more issues before merging.

File: pylint/checkers/variables.py
Comment (pdbsvmct) [reply pylint/checkers/variables.py:1338]: Don't we want to test for cases like `for i,j,k in {(a,b):1, (d,e): 2}.keys()`?

Comment (Neowizard) [reply pylint/checkers/variables.py:1338]: Don't we want to test for cases like `for i,j in {(a,b,c):1, (d,e,f): 2}.keys()` ?

Comment (mbyrnepr2) [reply pylint/checkers/variables.py:1338]: Yeah then we could also check for `nodes.DictKeys` perhaps?
I think it would be consistent with the other code above where we have a block of code specifically for `DictItems`: `if isinstance(inferred, astroid.objects.DictItems):` and now we would be explicitly dealing with these other situations `items()` and `keys()`.

Comment (Neowizard) [reply pylint/checkers/variables.py:1338]: Got it, I'll refactor the code to be explicit about the cases each branch covers

**Comment by Neowizard** [comment]
@mbyrnepr2, let me know if my latest commit is what you had in mind

**Comment by mbyrnepr2** [comment]
Yes looks great @Neowizard 👍

File: pylint/checkers/variables.py
Comment (mbyrnepr2) [inline pylint/checkers/variables.py:1345]:
```suggestion
        if isinstance(inferred, nodes.Dict):
            if isinstance(node.iter, nodes.Name):
                # If this a case of 'dict-items-missing-iter', we don't want to
                # report it as an 'unbalanced-dict-unpacking' as well
                # TODO (performance), merging both checks would streamline this
                if len(targets) == 2:
                    return

        else:
            is_stared_targets = any(
```
Perhaps a bit better like this since `inferred` is one of Dict, DictKeys, DictValues, DictItems.

Comment (Neowizard) [reply pylint/checkers/variables.py]: Yeah, that reads better, and passes tests. Committed.

File: pylint/checkers/variables.py
Comment (mbyrnepr2) [reply pylint/checkers/variables.py:3006]: Maybe we can omit this or consider it later if it is a bit niche.

**Review by mbyrnepr2 — APPROVED** [review]

**Review by mbyrnepr2 — APPROVED** [review]
👍

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas How far are we to releasing `3.x`? Do we really still need to backport?

**Comment by Pierre-Sassoulas** [comment]
I would say still pretty far (we did not start #3512 or anything in https://github.com/pylint-dev/pylint/issues?q=is%3Aopen+is%3Aissue+label%3A%22Blocker+%F0%9F%99%85%22 yet). But if something is really hard to backport we're not obligated to do it (hard defined as "massive cherry-pick conflict" right now but it could evolve to "any cherry-pick conflicts" if we want)

**Comment by Pierre-Sassoulas** [comment]
(Added the "needs backport" label in order to remember that at 2.17.6 release time, it does not mean that I think we NEED to backport :) )

**Comment by DanielNoord** [comment]
Thanks! I'm fine with backporting as well, but can't check out `pylint` on the laptop I'm currently on.

**Pull request merged**
